main.rs

```diff
@@ -0,0 +1,382 @@
+#[derive(Clone, PartialEq, ::prost::Message)]
+pub struct Fee {
+    /// amount is the amount of coins to be paid as a fee
+    #[prost(message, repeated, tag = "1")]
+    pub amount: ::std::vec::Vec<Coin>,
+    /// gas_limit is the maximum gas that can be used in transaction processing
+    /// before an out of gas error occurs
+    #[prost(uint64, tag = "2")]
+    pub gas_limit: u64,
+}
+
+/// SignMode represents a signing mode with its own security guarantees.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, ::prost::Enumeration)]
+#[repr(i32)]
+pub enum SignMode {
+    /// SIGN_MODE_UNSPECIFIED specifies an unknown signing mode and will be rejected
+    Unspecified = 0,
+    /// SIGN_MODE_DIRECT specifies a signing mode which uses SignDoc and is verified
+    /// with raw bytes from Tx
+    Direct = 1,
+    /// SIGN_MODE_TEXTUAL is a future signing mode that will verify some human-readable
+    /// textual representation on top of the binary representation from SIGN_MODE_DIRECT
+    Textual = 2,
+    /// SIGN_MODE_LEGACY_AMINO_JSON is a backwards compatibility mode which uses
+    /// Amino JSON and will be removed in the future
+    LegacyAminoJson = 127,
+}
+
+/// SignerInfo describes the public key and signing mode of a single top-level signer.
+#[derive(Clone, PartialEq, ::prost::Message)]
+pub struct SignerInfo {
+    /// public_key is the public key of the signer. It is optional for accounts
+    /// that already exist in state. If unset, the verifier can use the required \
+    /// signer address for this position and lookup the public key.
+    #[prost(message, optional, tag = "1")]
+    pub public_key: ::std::option::Option<PK>,
+    /// mode_info describes the signing mode of the signer and is a nested
+    /// structure to support nested multisig pubkey's
+    #[prost(message, optional, tag = "2")]
+    pub mode_info: ::std::option::Option<ModeInfo>,
+    /// sequence is the sequence of the account, which describes the
+    /// number of committed transactions signed by a given address. It is used to prevent
+    /// replay attacks.
+    #[prost(uint64, tag = "3")]
+    pub sequence: u64,
+}
+/// ModeInfo describes the signing mode of a single or nested multisig signer.
+#[derive(Clone, PartialEq, ::prost::Message)]
+pub struct ModeInfo {
+    /// sum is the oneof that specifies whether this represents a single or nested
+    /// multisig signer
+    #[prost(oneof = "Sum", tags = "1, 2")]
+    pub sum: ::std::option::Option<Sum>,
+}
+
+/// Single is the mode info for a single signer. It is structured as a message
+/// to allow for additional fields such as locale for SIGN_MODE_TEXTUAL in the future
+#[derive(Clone, PartialEq, ::prost::Message)]
+pub struct Single {
+    /// mode is the signing mode of the single signer
+    #[prost(enumeration = "SignMode", tag = "1")]
+    pub mode: i32,
+}
+
+/// CompactBitArray is an implementation of a space efficient bit array.
```

Comment:
Was this comment and accompanying code copied from the go codebase?

Reply:
I think this PR might be closed. This code was just added as a reference for the tx signing. It will be properly implemented when tendermint-rs upgrade lands on ibc-rs.